File: frontend/src/api/checkin.ts
/**
 * Checkin API endpoints
 * Handles daily check-in operations
 */

import { apiClient } from './client'

export interface CheckinRecord {
  id: number
  checkin_date: string
  streak: number
  base_amount: number
  milestone_amount: number
  total_amount: number
  balance_type: string
  created_at: string
}

export interface CheckinMilestone {
  days: number
  amount: number
  balance_type: string
  expiry_days: number
}

export interface CheckinConfig {
  mode: string
  fixed_amount: number
  random_min: number
  random_max: number
  balance_type: string
  expiry_days: number
  milestones: CheckinMilestone[]
}

export interface CheckinStatus {
  enabled: boolean
  checked_in_today: boolean
  current_streak: number
  total_checkins: number
  last_checkin_date: string
  config?: CheckinConfig
}

export interface CheckinCalendar {
  year: number
  month: number
  records: CheckinRecord[]
}

/**
 * Perform daily check-in
 */
export async function checkin(): Promise<CheckinRecord> {
  const { data } = await apiClient.post<CheckinRecord>('/user/checkin')
  return data
}

/**
 * Get check-in status for current user
 */
export async function getStatus(): Promise<CheckinStatus> {
  const { data } = await apiClient.get<CheckinStatus>('/user/checkin/status')
  return data
}

/**
 * Get monthly check-in calendar
 */
export async function getCalendar(year: number, month: number): Promise<CheckinCalendar> {
  const { data } = await apiClient.get<CheckinCalendar>('/user/checkin/calendar', {
    params: { year, month }
  })
  return data
}

export const checkinAPI = {
  checkin,
  getStatus,
  getCalendar,
}

File: backend/internal/service/checkin.go
package service

import (
	"context"
	"time"
)

// CheckinRecord 签到记录
type CheckinRecord struct {
	ID              int64
	UserID          int64
	CheckinDate     time.Time // 签到日期 (date only)
	Streak          int       // 连续签到天数（含当天）
	BaseAmount      float64   // 基础奖励
	MilestoneAmount float64   // 里程碑奖励
	TotalAmount     float64   // 总奖励
	BalanceType     string    // permanent / expirable
	CreatedAt       time.Time
}

// CheckinMilestone 里程碑规则
type CheckinMilestone struct {
	Days        int     `json:"days"`         // 连续签到满N天
	Amount      float64 `json:"amount"`       // 奖励金额
	BalanceType string  `json:"balance_type"` // permanent / expirable
	ExpiryDays  int     `json:"expiry_days"`  // 有效期天数（仅 expirable 生效）
}

// CheckinConfig 签到配置
type CheckinConfig struct {
	Enabled     bool                `json:"enabled"`
	Mode        string              `json:"mode"`         // fixed / random
	FixedAmount float64             `json:"fixed_amount"`
	RandomMin   float64             `json:"random_min"`
	RandomMax   float64             `json:"random_max"`
	BalanceType string              `json:"balance_type"` // permanent / expirable
	ExpiryDays  int                 `json:"expiry_days"`  // 有效期天数（仅 expirable 生效）
	Milestones  []CheckinMilestone  `json:"milestones"`
}

// CheckinStatus 用户签到状态
type CheckinStatus struct {
	CheckedInToday bool   `json:"checked_in_today"`
	CurrentStreak  int    `json:"current_streak"`
	TotalCheckins  int64  `json:"total_checkins"`
	LastCheckinDate string `json:"last_checkin_date,omitempty"`
}

// CheckinRepository 签到仓储接口
type CheckinRepository interface {
	// Create 创建签到记录
	Create(ctx context.Context, record *CheckinRecord) error
	// GetByUserAndDate 按用户和日期查询签到记录
	GetByUserAndDate(ctx context.Context, userID int64, date time.Time) (*CheckinRecord, error)
	// GetLatestByUser 获取用户最近一条签到记录
	GetLatestByUser(ctx context.Context, userID int64) (*CheckinRecord, error)
	// ListByUser 查询用户签到历史（分页）
	ListByUser(ctx context.Context, userID int64, offset, limit int) ([]*CheckinRecord, int64, error)
	// CountByUser 统计用户总签到次数
	CountByUser(ctx context.Context, userID int64) (int64, error)
	// GetMonthlyRecords 获取用户某月的签到记录（日历展示用）
	GetMonthlyRecords(ctx context.Context, userID int64, year int, month int) ([]*CheckinRecord, error)
}
